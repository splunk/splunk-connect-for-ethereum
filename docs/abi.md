# ABI Decoding

Splunk Connect for Ethereum features decoding of function calls and event logs in Ethereum transactions based on [Solidity ABI](https://solidity.readthedocs.io/en/develop/abi-spec.html) information.

In addition to the binary data of transactions, the following information is extracted:

-   Function or event name
-   Parameter values and data types
-   Parameter names
-   Contract name

This information can be used to easily search for semantic aspects of smart contracts and it enables powerful and efficient aggregations and analytics.

## Supplying ABI Definitions

You can supply ABI definition files to ethlogger by specifying a directory in which ethlogger should search for them. This can be done by specifying the `--abi-dir` [CLI flag](./cli.md), or by providing the corresponding [environment variable](./cli.md#environment-variables) `ABI_DIR`. For more fine-grained control, you can configure the [ABI repository in ethlogger configuration file](./configuration.md#abirepository).

Each ABI file corresponds to a smart contract and contains all function and event signatures, and optionally a contract name and address at which the contract is deployed.

Matching ABI defintions to transactions follows the following priority order:

-   Contract address (if contract address from ABI file matches address from transaction)
-   Contract fingerprint (if contract fingerprint matches fingerprint derived from contract code)
-   Anonymous signature

### File Formats

Files in the ABI directory are expected to contain [ABI JSON](https://solidity.readthedocs.io/en/develop/abi-spec.html#abi-json), either as plain list of ABI items, or truffle build files.

#### List of ABI Items

ABI files may contain an array of ABI items. In this case the contract name is derived from the file name (by removing the file extension) and no contract address is associated.

**Example**:

```json
[
    {
        "type": "function",
        "name": "transfer",
        "inputs": [
            {
                "name": "recipient",
                "type": "address"
            },
            {
                "name": "amount",
                "type": "uint256"
            }
        ]
    },
    {
        "type": "event",
        "name": "TransferApproved",
        "inputs": [
            {
                "name": "recipient",
                "type": "address"
            },
            {
                "name": "amount",
                "type": "uint256"
            }
        ]
    }
]
```

#### Truffle Build Files

Truffle build files can have information about which network and contract address a contract has been deployed to. This information is used by ethlogger to fast-track matching an address it encounters on the blockchain to the ABI definition. Furthermore, truffle build files can specify a contract name; if not present the file name will be used as contract name. Files are expected to have a property `abi` containing an array of ABI items, all other information is optional.

```jsonc
{
    "contractName": "My Contract",
    "abi": [
        /* ... ABI items */
    ],
    "networks": {
        "1": {
            "address": "0x123..."
        }
    }
}
```

## Contract Fingerprinting

One challenge that can occur when decoding ABI information is overlapping signatures. Imagine 2 smart contracts both exposing a function with the signature `transfer(address,uint256)`. For one contract the second argument may refer to an amount of currency being transferred (eg. `transfer(to address, amount uint256)`) and for another contract the second argument refers to the ID of a non-fungible item (eg. `transfer(to address, tokenId uint256)`). Given the nature of ABI encoding, both functions end up having the same signature and it's ambiguous for the ABI decoder which one to choose, unless the smart contract can be associated with the ABI definition (if the ABI definition contains the contract address).

It would be nice, to be able to decode any instance of a smart contract, like an ERC20 and not having to enumerate all the address the contract has been deployed to.

An attempt to solve this problem is "contract fingerprinting". A fingerprint is a hash of all function and event signatures from a contract ABI. We can also derive the fingerprint from the EVM bytecode by extracting all signature hashes, finding matching signatures for them, and computing the fingerprint from it, and then comparing that against the fingerprint computed from the ABI defintion. There is still the possibility of invalid matches, especially for contracts that only have very few functions/events, but with multiple functions the chances are much lower.

Fingerprinting is enabled by default and can be disabled in the [ethlogger configuration](./configuration.md#abirepository).

## Anonymous ABI Decoding

For cases where we can find a match with between a deployed contract and a supplied ABI definition, but do have a match for the function or event signature hash, we can emit the reduced amount of information we can derive from the signature alone (the function/event name, the parameter data types and values, but not the parameter names or the contract name).

If anonymous ABI encoding is enabled (which it is by default), then this reduced amount of information is emitted. In addition to the ABIs supplied by the user, ethlogger ships with a standard set of function and event signatures that are compiled from from external sources:

-   https://github.com/MrLuit/evm
-   https://www.4byte.directory/

Compiled lists can be found here:

-   Function signatures: [function_signatures.txt](../data/function_signatures.txt)
-   Event signatures: [event_signatures.txt](../data/event_signatures.txt)

Those anonymous signature are loaded by default.
