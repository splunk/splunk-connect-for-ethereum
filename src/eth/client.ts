import { sleep } from '../utils/async';
import { createModuleDebug } from '../utils/debug';
import { checkError, createJsonRpcPayload, JsonRpcError, JsonRpcRequest } from './jsonrpc';
import { EthRequest } from './requests';
import { EthereumTransport } from './transport';

const { debug, error } = createModuleDebug('eth:client');

interface BatchReq<P extends any[] = any[], R = any> {
    request: EthRequest<P, R>;
    callback: (error: Error | null, result: R) => void;
}

export async function executeBatchRequest(batch: BatchReq[], transport: EthereumTransport) {
    debug('Processing batch of %d JSON RPC requests', batch.length);
    try {
        const items = new Map<number, BatchReq>();
        const reqs: JsonRpcRequest[] = [];
        for (const batchItem of batch) {
            const req = createJsonRpcPayload(batchItem.request.method, batchItem.request.params);
            reqs.push(req);
            items.set(req.id, batchItem);
        }
        const results = await transport.sendBatch(reqs);
        for (const result of results) {
            const batchItem = items.get(result.id);
            if (batchItem == null) {
                error(`Found unassociated batch item in batch response`);
                continue;
            }
            batchItem.callback(null, result);
            items.delete(result.id);
        }
        if (items.size > 0) {
            error(`Unprocessed batch request after receiving results`);
            for (const unprocessed of items.values()) {
                unprocessed.callback(new Error('Result missing from batch response'), null);
            }
        }
    } catch (e) {
        batch.forEach(({ callback }) => callback(e, null));
    }
}

export class EthereumClient {
    constructor(protected transport: EthereumTransport) {}

    async request<P extends any[], R>(req: EthRequest<P, R>): Promise<R> {
        const payload = createJsonRpcPayload(req.method, req.params);
        const res = await this.transport.send(payload);
        if (payload.id !== res.id) {
            throw new JsonRpcError(`JSON RPC Response ID mismatch. Expected ${payload.id} but got ${res.id}`);
        }
        checkError(res);
        return typeof req.response === 'function' ? req.response(res) : res.result;
    }

    async requestBatch<P extends any[], R>(reqs: EthRequest<P, R>[]): Promise<R[]> {
        const batchReqs: BatchReq[] = [];
        const promises = [];
        for (const request of reqs) {
            promises.push(
                new Promise<R>((resolve, reject) => {
                    batchReqs.push({
                        request,
                        callback: (e, result) => {
                            if (e) {
                                reject(e);
                            }
                            try {
                                checkError(result);
                                resolve(
                                    typeof request.response === 'function' ? request.response(result) : result.result
                                );
                            } catch (e) {
                                debug(`JSON RPC request method: %s failed`, request.method, e);
                                reject(e);
                            }
                        },
                    });
                })
            );
        }
        await sleep(0);
        await executeBatchRequest(batchReqs, this.transport);
        return await Promise.all(promises);
    }
}

export interface BatchedEthereumClientConfig {
    maxBatchTime: number;
    maxBatchSize: number;
}

export class BatchedEthereumClient extends EthereumClient {
    private config: BatchedEthereumClientConfig;
    private queue: BatchReq<any[], any>[] = [];
    private flushTimer: NodeJS.Timer | null = null;
    constructor(transport: EthereumTransport, config: BatchedEthereumClientConfig) {
        super(transport);
        this.config = config;
    }

    request<P extends any[], R>(
        req: EthRequest<P, R>,
        { immediate = false }: { immediate?: boolean } = {}
    ): Promise<R> {
        if (immediate) {
            return super.request(req);
        }

        return new Promise((resolve, reject) => {
            this.queue.push({
                request: req,
                callback: (e, result) => {
                    if (e) {
                        reject(e);
                    }
                    try {
                        checkError(result);
                        resolve(typeof req.response === 'function' ? req.response(result) : result.result);
                    } catch (e) {
                        debug(`JSON RPC request method: %s(%o) failed`, req.method, req.params, e);
                        reject(e);
                    }
                },
            });
            this.scheduleFlush();
        });
    }

    private processQueue = async () => {
        if (this.flushTimer != null) {
            clearTimeout(this.flushTimer);
            this.flushTimer = null;
        }
        if (this.queue.length === 0) {
            return;
        }
        const batch = this.queue;
        this.queue = [];
        return await executeBatchRequest(batch, this.transport);
    };

    private scheduleFlush() {
        if (this.queue.length >= this.config.maxBatchSize) {
            this.processQueue();
            return;
        }
        if (this.flushTimer == null) {
            this.flushTimer = setTimeout(this.processQueue, this.config.maxBatchTime);
        }
    }
}
