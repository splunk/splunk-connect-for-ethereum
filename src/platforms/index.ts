import { OutputMessage } from '../output';
import { EthereumClient } from '../eth/client';

/** Generic interface to capture platform-specific metrics and data from nodes */
export interface NodePlatformAdapter {
    readonly name: string;
    readonly fullVersion: string;
    readonly enode: string | null;

    initialize?(ethClient: EthereumClient): Promise<void>;

    captureNodeStats(ethClient: EthereumClient): Promise<OutputMessage[]>;
}
