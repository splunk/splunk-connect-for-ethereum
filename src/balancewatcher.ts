import { createModuleDebug } from './utils/debug';
import { ManagedResource } from './utils/resource';
import { EthereumClient } from './eth/client';
import { Output, OutputMessage } from './output';
import { BalanceWatcherConfig } from './config';
import { NodePlatformAdapter } from './platforms';
import { ABORT, AbortHandle } from './utils/abort';
import { linearBackoff, resolveWaitTime, retry, WaitTime } from './utils/retry';
import { Cache } from './utils/cache';
import { ContractInfo } from './abi/contract';
import { AggregateMetric } from './utils/stats';
import { blockNumber, ethBalance, ethCall, getBlock, getTransactionReceipt } from './eth/requests';
import { BlockRange, blockRangeSize, blockRangeToArray, chunkedBlockRanges, serializeBlockRange } from './blockrange';
import { parallel, sleep } from './utils/async';
import { RawBlockResponse, RawTransactionResponse } from './eth/responses';
import { bigIntToNumber } from './utils/bn';
import { formatBlock, formatHexToFloatingPoint } from './format';
import { parseBlockTime } from './blockwatcher';
import { BalanceMessage, FormattedBlock } from './msgs';
import { sha3 } from './abi/wasm';
import { Checkpoint } from './state';

const { debug, info, warn, error, trace } = createModuleDebug('balancewatcher');

const transferEventSignature = sha3('Transfer(address,address,uint256)')!;

const initialCounters = {
    blocksProcessed: 0,
    transfersProcessed: 0,
};

/**
 * BalanceWatcher will watch the activity of a contract on chain, and keep track of all accounts transferring tokens.
 */
export class BalanceWatcher implements ManagedResource {
    private active: boolean = true;
    private ethClient: EthereumClient;
    private checkpoint: Checkpoint;
    private output: Output;
    private config: BalanceWatcherConfig;
    private nodePlatform: NodePlatformAdapter;
    private abortHandle = new AbortHandle();
    private endCallbacks: Array<() => void> = [];
    private waitAfterFailure: WaitTime;
    private chunkQueueMaxSize: number;
    private counters = { ...initialCounters };
    private aggregates = {
        blockProcessTime: new AggregateMetric(),
        txProcessTime: new AggregateMetric(),
        eventProcessTime: new AggregateMetric(),
    };

    constructor({
        ethClient,
        checkpoint,
        output,
        config,
        waitAfterFailure = linearBackoff({ min: 0, step: 2500, max: 120_000 }),
        chunkQueueMaxSize = 1000,
        nodePlatform,
    }: {
        ethClient: EthereumClient;
        output: Output;
        checkpoint: Checkpoint;
        config: BalanceWatcherConfig;
        waitAfterFailure?: WaitTime;
        chunkQueueMaxSize?: number;
        contractInfoCache?: Cache<string, Promise<ContractInfo>>;
        nodePlatform: NodePlatformAdapter;
    }) {
        this.ethClient = ethClient;
        this.checkpoint = checkpoint;
        this.output = output;
        this.config = config;
        this.waitAfterFailure = waitAfterFailure;
        this.chunkQueueMaxSize = chunkQueueMaxSize;
        this.nodePlatform = nodePlatform;
    }

    async start(): Promise<void> {
        debug('Starting balances watcher for %s', this.config.contractAddress);
        const { ethClient, checkpoint } = this;

        if (checkpoint.isEmpty()) {
            if (typeof this.config.startAt === 'number') {
                if (this.config.startAt < 0) {
                    const latestBlockNumber = await ethClient.request(blockNumber());
                    checkpoint.setInitialBlockNumber(Math.max(0, latestBlockNumber + this.config.startAt));
                } else {
                    checkpoint.setInitialBlockNumber(this.config.startAt);
                }
            } else if (this.config.startAt === 'genesis') {
                checkpoint.setInitialBlockNumber(0);
            } else if (this.config.startAt === 'latest') {
                const latestBlockNumber = await ethClient.request(blockNumber());
                checkpoint.setInitialBlockNumber(latestBlockNumber);
            } else {
                throw new Error(`Invalid start block: ${JSON.stringify(this.config.startAt)}`);
            }
            info(
                'Determined initial block number: %d from configured value %o',
                checkpoint.getinitialBlockNumber(),
                this.config.startAt
            );
        }

        let failures = 0;
        while (this.active) {
            try {
                let latestBlockNumber = await ethClient.request(blockNumber());
                debug('Received latest block number: %d', latestBlockNumber);
                const reachedEnd = this.config.endAt !== undefined && latestBlockNumber > this.config.endAt;
                if (reachedEnd) {
                    latestBlockNumber = this.config.endAt as number;
                    debug('Processing up to end of range %d', latestBlockNumber);
                }
                const todo = checkpoint.getIncompleteRanges(latestBlockNumber);
                debug('Found %d block ranges to process', todo.length);
                if (todo.length) {
                    info(
                        'Latest block number is %d - processing remaining %d blocks',
                        latestBlockNumber,
                        todo.map(blockRangeSize).reduce((a, b) => a + b, 0)
                    );
                }
                if (reachedEnd && todo.length == 0) {
                    info('Completed collecting blocks from %o to %d', this.config.startAt, this.config.endAt);
                    this.active = false;
                }

                for (const range of todo) {
                    if (!this.active) {
                        throw ABORT;
                    }
                    await parallel(
                        chunkedBlockRanges(range, this.config.blocksMaxChunkSize, this.chunkQueueMaxSize).map(
                            chunk => async () => {
                                if (!this.active) {
                                    return;
                                }
                                await retry(() => this.processChunk(chunk), {
                                    attempts: 100,
                                    waitBetween: this.waitAfterFailure,
                                    taskName: `block range ${serializeBlockRange(chunk)}`,
                                    abortHandle: this.abortHandle,
                                    warnOnError: true,
                                    onRetry: attempt =>
                                        warn(
                                            'Retrying to process block range %s (attempt %d)',
                                            serializeBlockRange(chunk),
                                            attempt
                                        ),
                                });
                            }
                        ),
                        { maxConcurrent: this.config.maxParallelChunks, abortHandle: this.abortHandle }
                    );
                    failures = 0;
                }
            } catch (e) {
                if (e === ABORT) {
                    break;
                }
                error('Error in block watcher polling loop', e);
                failures++;
                const waitTime = resolveWaitTime(this.waitAfterFailure, failures);
                if (waitTime > 0) {
                    warn('Waiting for %d ms after %d consecutive failures', waitTime, failures);
                    await this.abortHandle.race(sleep(waitTime));
                }
            }
            if (this.active) {
                try {
                    await this.abortHandle.race(sleep(this.config.pollInterval));
                } catch (e) {
                    if (e === ABORT) {
                        break;
                    } else {
                        warn('Unexpected error while waiting for next polling loop iteration', e);
                    }
                }
            }
        }
        if (this.endCallbacks != null) {
            this.endCallbacks.forEach(cb => cb());
        }
        info('Balances watcher stopped');
    }

    async processChunk(chunk: BlockRange) {
        const startTime = Date.now();
        info('Processing chunk %s', serializeBlockRange(chunk));

        debug('Requesting block range', chunk);
        const blockRequestStart = Date.now();
        const blocks = await this.ethClient
            .requestBatch(
                blockRangeToArray(chunk)
                    .filter(blockNumber => this.checkpoint.isIncomplete(blockNumber))
                    .map(blockNumber => getBlock(blockNumber))
            )
            .catch(e =>
                Promise.reject(new Error(`Failed to request batch of blocks ${serializeBlockRange(chunk)}: ${e}`))
            );
        debug('Received %d blocks in %d ms', blocks.length, Date.now() - blockRequestStart);
        for (const block of blocks) {
            await this.processBlock(block);
        }
        info(
            'Completed %d blocks of chunk %s in %d ms',
            blocks.length,
            serializeBlockRange(chunk),
            Date.now() - startTime
        );
    }

    async processBlock(block: RawBlockResponse) {
        if (block.number != null && !this.checkpoint.isIncomplete(bigIntToNumber(block.number))) {
            warn('Skipping processing of block %d since it is marked complete in our checkpoint');
            return;
        }
        const startTime = Date.now();
        const formattedBlock = formatBlock(block);
        if (formattedBlock.number == null) {
            debug('Ignoring block %s without number', block.hash);
            return;
        }
        const blockTime = parseBlockTime(formattedBlock.timestamp);

        if (!this.active) {
            return;
        }
        const outputMessages = await this.abortHandle.race(
            Promise.all(block.transactions.map(tx => this.processTransaction(tx, blockTime, formattedBlock)))
        );
        outputMessages.forEach(msgs => msgs.forEach(msg => this.output.write(msg)));
        this.checkpoint.markBlockComplete(formattedBlock.number);
        this.counters.blocksProcessed++;
        this.aggregates.blockProcessTime.push(Date.now() - startTime);
    }

    public async shutdown() {
        info('Shutting down balances watcher for %s', this.config.contractAddress);
        this.active = false;
        this.abortHandle.abort();
        await new Promise<void>(resolve => {
            this.endCallbacks.push(resolve);
        });
    }

    public flushStats() {
        const stats = {
            ...this.counters,
            ...this.aggregates.blockProcessTime.flush('blockProcessTime'),
            ...this.aggregates.txProcessTime.flush('txProcessTime'),
            ...this.aggregates.eventProcessTime.flush('eventProcessTime'),
            abortHandles: this.abortHandle.size,
        };
        this.counters = { ...initialCounters };
        return stats;
    }

    async processTransaction(
        rawTx: RawTransactionResponse | string,
        blockTime: number,
        formattedBlock: FormattedBlock
    ): Promise<OutputMessage[]> {
        if (!this.active) {
            return [];
        }
        if (typeof rawTx === 'string') {
            warn('Received raw transaction as string from block %d', formattedBlock.number);
            return [];
        }
        const startTime = Date.now();
        trace('Processing transaction %s from block %d', rawTx.hash, formattedBlock.number);

        const receipt = await this.ethClient.request(getTransactionReceipt(rawTx.hash));

        let outputMessages = Array<OutputMessage>();
        if (receipt != null && receipt.logs != null) {
            const addresses = receipt.logs
                ?.map(log => {
                    if (this.config.contractAddress == log.address) {
                        if (log.topics[0] == transferEventSignature && log.topics.length == 3) {
                            const from = '0x' + log.topics[1].substr(26);
                            const to = '0x' + log.topics[2].substr(26);
                            return [from, to];
                        }
                    }
                })
                .flat()
                .filter(a => a !== undefined);
            if (addresses.length > 0) {
                this.counters.transfersProcessed++;
                outputMessages = await this.abortHandle.race(
                    Promise.all(addresses.map(addr => this.getBalance(addr, formattedBlock, blockTime)))
                );
            }
        }

        this.aggregates.txProcessTime.push(Date.now() - startTime);
        trace('Completed processing transaction %s in %d ms', rawTx.hash, Date.now() - startTime);
        return outputMessages;
    }

    async getBalance(address: string, formattedBlock: FormattedBlock, blockTime: number): Promise<BalanceMessage> {
        const response = await this.ethClient.request(
            ethCall(
                this.config.contractAddress,
                'balanceOf(address)',
                ['0'.repeat(24) + address.substr(2)],
                formattedBlock.number!
            )
        );
        const balance = formatHexToFloatingPoint(response, this.config.decimals);

        let ethereumBalance = undefined;
        if (this.config.logEthBalance) {
            const ethereumBalanceHex = await this.ethClient.request(ethBalance(address, formattedBlock.number!));
            ethereumBalance = formatHexToFloatingPoint(ethereumBalanceHex, 18);
        }

        return {
            body: {
                contract: this.config.contractAddress,
                blockHash: formattedBlock.hash!,
                blockNumber: formattedBlock.number!,
                account: address,
                balance: balance,
                ethBalance: ethereumBalance,
            },
            time: blockTime,
            type: 'balance',
        };
    }
}
