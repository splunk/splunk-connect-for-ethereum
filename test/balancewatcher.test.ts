import { join } from 'path';
import { ContractInfo } from '../src/abi/contract';
import { State } from '../src/state';
import { BatchedEthereumClient } from '../src/eth/client';
import { HttpTransport } from '../src/eth/http';
import { withRecorder } from '../src/eth/recorder';
import { suppressDebugLogging } from '../src/utils/debug';
import LRUCache from '../src/utils/lru';
import { TestOutput } from './testoutput';
import { MOCK_NODE_ADAPTER } from '../src/platforms/mock';
import { BalanceWatcher } from '../lib/balancewatcher';

let logHandle: any;
beforeEach(() => {
    logHandle = suppressDebugLogging();
});
afterEach(() => {
    logHandle.restore();
});

test('balancewatcher', async () => {
    await withRecorder(
        new HttpTransport('https://dai.poa.network', {}),
        {
            name: 'xdai-balancewatcher',
            storageDir: join(__dirname, './fixtures/recorded'),
            replay: true,
        },
        async transport => {
            const ethClient = new BatchedEthereumClient(transport, { maxBatchSize: 100, maxBatchTime: 0 });
            const state = new State({
                path: join(__dirname, '../tmp/tmpcheckpoint.json'),
                saveInterval: 10000,
            });
            const checkpoint = state.getCheckpoint('main');
            checkpoint.setInitialBlockNumber(123);
            const output = new TestOutput();
            const contractInfoCache = new LRUCache<string, Promise<ContractInfo>>({ maxSize: 100 });
            const balanceWatcher = new BalanceWatcher({
                checkpoint,
                config: {
                    contractAddress: '0xe91d153e0b41518a2ce8dd3d7944fa863463a97d',
                    decimals: 18,
                    enabled: true,
                    blocksMaxChunkSize: 1,
                    pollInterval: 1,
                    maxParallelChunks: 1,
                    startAt: 'latest',
                    retryWaitTime: 10,
                    logEthBalance: true,
                },
                ethClient,
                output,
                contractInfoCache,
                waitAfterFailure: 1,
                nodePlatform: MOCK_NODE_ADAPTER,
            });

            await balanceWatcher.processChunk({ from: 19588262, to: 19588292 });

            expect(output.messages).toMatchSnapshot();
        }
    );
}, 15000);
