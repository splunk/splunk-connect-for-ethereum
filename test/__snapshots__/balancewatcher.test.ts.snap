// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`balancewatcher 1`] = `
Array [
  Object {
    "body": Object {
      "account": "0x8c0c36c85192204c8d782f763ff5a30f5ba0192f",
      "balance": "2658482.024187534230024041",
      "blockHash": "0x15d73413b167b8fe2ac9746ce03b74caf20c712d2b0144ffec89a4b02b31fa69",
      "blockNumber": 19588263,
      "contract": "0xe91d153e0b41518a2ce8dd3d7944fa863463a97d",
      "ethBalance": "0.000000000000000000",
    },
    "time": 1639552555000,
    "type": "balance",
  },
  Object {
    "body": Object {
      "account": "0x9008d19f58aabd9ed0d60971565aa8510560ab41",
      "balance": "396.601497742790230302",
      "blockHash": "0x15d73413b167b8fe2ac9746ce03b74caf20c712d2b0144ffec89a4b02b31fa69",
      "blockNumber": 19588263,
      "contract": "0xe91d153e0b41518a2ce8dd3d7944fa863463a97d",
      "ethBalance": "0.000000000000000014",
    },
    "time": 1639552555000,
    "type": "balance",
  },
  Object {
    "body": Object {
      "account": "0x1865d5445010e0baf8be2eb410d3eae4a68683c2",
      "balance": "1086971.038410192057465927",
      "blockHash": "0x15d73413b167b8fe2ac9746ce03b74caf20c712d2b0144ffec89a4b02b31fa69",
      "blockNumber": 19588263,
      "contract": "0xe91d153e0b41518a2ce8dd3d7944fa863463a97d",
      "ethBalance": "0.000000000000000000",
    },
    "time": 1639552555000,
    "type": "balance",
  },
  Object {
    "body": Object {
      "account": "0x9008d19f58aabd9ed0d60971565aa8510560ab41",
      "balance": "396.601497742790230302",
      "blockHash": "0x15d73413b167b8fe2ac9746ce03b74caf20c712d2b0144ffec89a4b02b31fa69",
      "blockNumber": 19588263,
      "contract": "0xe91d153e0b41518a2ce8dd3d7944fa863463a97d",
      "ethBalance": "0.000000000000000014",
    },
    "time": 1639552555000,
    "type": "balance",
  },
  Object {
    "body": Object {
      "account": "0xe5759714998e8b50a33c7333c04c2d02e5dce77f",
      "balance": "0.000000000000000000",
      "blockHash": "0xb1fb0b96af1e305513c8cb992567b899ac376cf0081760bd79abe15a5e686b27",
      "blockNumber": 19588276,
      "contract": "0xe91d153e0b41518a2ce8dd3d7944fa863463a97d",
      "ethBalance": "0.000000000000000000",
    },
    "time": 1639552620000,
    "type": "balance",
  },
  Object {
    "body": Object {
      "account": "0x52eeff79eeceaf8162b4e5fa383dcfea4d78dc98",
      "balance": "162055.903714243311408350",
      "blockHash": "0xb1fb0b96af1e305513c8cb992567b899ac376cf0081760bd79abe15a5e686b27",
      "blockNumber": 19588276,
      "contract": "0xe91d153e0b41518a2ce8dd3d7944fa863463a97d",
      "ethBalance": "0.000000000000000000",
    },
    "time": 1639552620000,
    "type": "balance",
  },
  Object {
    "body": Object {
      "account": "0x24afdca4653042c6d08fb1a754b2535dacf6eb24",
      "balance": "35029.885008572241645008",
      "blockHash": "0xfb6ad9fae309517039adcdd9907584df3ca29374b4777f3fa7828151ef4b7735",
      "blockNumber": 19588278,
      "contract": "0xe91d153e0b41518a2ce8dd3d7944fa863463a97d",
      "ethBalance": "0.000000000000000000",
    },
    "time": 1639552630000,
    "type": "balance",
  },
  Object {
    "body": Object {
      "account": "0xe432da6e92f024c68ebd300b7191a8e2859e9039",
      "balance": "31.063523989322339355",
      "blockHash": "0xfb6ad9fae309517039adcdd9907584df3ca29374b4777f3fa7828151ef4b7735",
      "blockNumber": 19588278,
      "contract": "0xe91d153e0b41518a2ce8dd3d7944fa863463a97d",
      "ethBalance": "1.998500324866891436",
    },
    "time": 1639552630000,
    "type": "balance",
  },
  Object {
    "body": Object {
      "account": "0x1b02da8cb0d097eb8d57a175b88c7d8b47997506",
      "balance": "0.000000000000000000",
      "blockHash": "0x2fa25e426ea987e6fb991fa42b9fd226a7223f196b6954ee4ca5945133ce20d1",
      "blockNumber": 19588279,
      "contract": "0xe91d153e0b41518a2ce8dd3d7944fa863463a97d",
      "ethBalance": "0.000000000000000000",
    },
    "time": 1639552635000,
    "type": "balance",
  },
  Object {
    "body": Object {
      "account": "0xc704050a17af0caed763431b80e38e8d8ff15591",
      "balance": "453856.482894956902087461",
      "blockHash": "0x2fa25e426ea987e6fb991fa42b9fd226a7223f196b6954ee4ca5945133ce20d1",
      "blockNumber": 19588279,
      "contract": "0xe91d153e0b41518a2ce8dd3d7944fa863463a97d",
      "ethBalance": "0.000000000000000000",
    },
    "time": 1639552635000,
    "type": "balance",
  },
  Object {
    "body": Object {
      "account": "0x5805bb63e73ec272c74e210d280c05b41d719827",
      "balance": "16212.975608344552347985",
      "blockHash": "0x2fa25e426ea987e6fb991fa42b9fd226a7223f196b6954ee4ca5945133ce20d1",
      "blockNumber": 19588279,
      "contract": "0xe91d153e0b41518a2ce8dd3d7944fa863463a97d",
      "ethBalance": "0.000000000000000000",
    },
    "time": 1639552635000,
    "type": "balance",
  },
  Object {
    "body": Object {
      "account": "0x1c232f01118cb8b424793ae03f870aa7d0ac7f77",
      "balance": "0.000000000000000000",
      "blockHash": "0x2fa25e426ea987e6fb991fa42b9fd226a7223f196b6954ee4ca5945133ce20d1",
      "blockNumber": 19588279,
      "contract": "0xe91d153e0b41518a2ce8dd3d7944fa863463a97d",
      "ethBalance": "0.000000000000000000",
    },
    "time": 1639552635000,
    "type": "balance",
  },
  Object {
    "body": Object {
      "account": "0xa227c72a4055a9dc949cae24f54535fe890d3663",
      "balance": "2069834.816844046862756244",
      "blockHash": "0xe26eab1018c141e566522a1ed418f7926c9f21bed9919e38a9cc3ce313ab0969",
      "blockNumber": 19588285,
      "contract": "0xe91d153e0b41518a2ce8dd3d7944fa863463a97d",
      "ethBalance": "0.000000000000000000",
    },
    "time": 1639552665000,
    "type": "balance",
  },
  Object {
    "body": Object {
      "account": "0x1b02da8cb0d097eb8d57a175b88c7d8b47997506",
      "balance": "0.000000000000000000",
      "blockHash": "0xe26eab1018c141e566522a1ed418f7926c9f21bed9919e38a9cc3ce313ab0969",
      "blockNumber": 19588285,
      "contract": "0xe91d153e0b41518a2ce8dd3d7944fa863463a97d",
      "ethBalance": "0.000000000000000000",
    },
    "time": 1639552665000,
    "type": "balance",
  },
  Object {
    "body": Object {
      "account": "0x2b6ebd2502b8182da13f9c4a71d4ba8a7a3f7011",
      "balance": "0.000000000000000000",
      "blockHash": "0xdaded7f6fb9c717eac15f09e2e5873e4570467495b37351f56569818e0fe65f9",
      "blockNumber": 19588292,
      "contract": "0xe91d153e0b41518a2ce8dd3d7944fa863463a97d",
      "ethBalance": "0.260765518572927979",
    },
    "time": 1639552700000,
    "type": "balance",
  },
  Object {
    "body": Object {
      "account": "0xb59da143ba3197dd5ea3844f429d37a1c2b1afd8",
      "balance": "336.607138288936731520",
      "blockHash": "0xdaded7f6fb9c717eac15f09e2e5873e4570467495b37351f56569818e0fe65f9",
      "blockNumber": 19588292,
      "contract": "0xe91d153e0b41518a2ce8dd3d7944fa863463a97d",
      "ethBalance": "0.000000000000000000",
    },
    "time": 1639552700000,
    "type": "balance",
  },
]
`;
